token-swap/program/src/curve/base.rs

```diff
@@ -37,6 +60,47 @@ pub struct SwapCurve {
     pub calculator: Box<dyn CurveCalculator>,
 }
 
+impl SwapCurve {
+    /// Subtract fees and calculate how much destination token will be provided
+    /// given an amount of source token.
+    pub fn swap(
+        &self,
+        source_amount: u128,
+        swap_source_amount: u128,
+        swap_destination_amount: u128,
+        trade_direction: TradeDirection,
+        fees: &Fees,
```

Comment:
Two new parameters to watch for here -- the fees are done from the outside now, so it made more sense to just move this into the `SwapCurve` and have the calculator stay totally fee-free